### PR TITLE
feat: ts-cli support to export data

### PR DESCRIPTION
**Pull request by wtsclwq**
### What problem does this PR solve?

Issue Number: fix #241

### What is changed and how it works?

Refer to the infuxdb1.x `influx_inspect export [options]`.
Exports all tssp and wal files to line protocol. This output file can be imported via the ts-cli command.
usage:Multiple

1. Export specified or all retentions for **one database** 

````bash
ts-cli -export -datadir=${data_dir} -waldir=${wal_dir} -databases=db0 -retentions={rp0,rp1...} -start=${start_time} -end=${end_time} -compress=true/false -out=output.txt -lponly=true/false
````

2. Export **multiple databases**, this usage cannot specify the retentions.

````bash
ts-cli -export -datadir=${data_dir} -waldir=${wal_dir} -databases={db0,db1...} -start=${start_time} -end=${end_time} -compress=true/false -out=output.txt -lponly=true/false
````

3. Export **all databases**,  this usage cannot specify the retentions.

```bash
ts-cli -export -datadir=${data_dir} -waldir=${wal_dir} -start=${start_time} -end=${end_time} -compress=true/false -out=output.txt -lponly=true/false

```

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

1. Learn from the test methods and data in `hard_test.go`, organize the directory of temp shard into a form of actual use, to the write data and export , and verify whether theyt are the same. 

2. Use the [[example data](https://docs.opengemini.org/zh/guide/geminiql/sample_data.html)](https://docs.opengemini.org/zh/guide/geminiql/sample_data.html) in the document to import and export it using ts-cli import, and check if the data is consistent. After clearing the database, use the exported file to re-import it, check whether it is consistent with the example data, repeat multiple times

- [ ] Test A
- [ ] Test B
- [x] Test cases to be added
- [ ] No code

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/ts-cli/geminicli/export_test.go [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+		"test_db2_rp2": "/tmp/og_test/wal/wal/test_db2/0/test_db2_rp2",
+	},
+	rpToIndexDirMap: map[string]string{
+		"test_db2_rp1": "/tmp/og_test/data/data/test_db2/0/test_db2_rp1/index",
+		"test_db2_rp2": "/tmp/og_test/data/data/test_db2/0/test_db2_rp2/index",
+	},
+}
+
+func createTestDirs() {
+	err := os.MkdirAll("/tmp/og_test/data/data/test_db1/0/test_db1_rp1/1_1568592000000000000_1569196800000000000_1", os.ModePerm)
```
Comment (shilinlee) [inline app/ts-cli/geminicli/export_test.go:65]: use `tmpDir := t.TempDir()` to generate temp directory.

File: app/ts-cli/geminicli/export_test.go [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+				}
+			}
+
+			copyRs = append(copyRs, tmpPointRow)
+		}
+	}
+	return copyRs
+}
+
+func removeTestDirs() {
```
Comment (shilinlee) [inline app/ts-cli/geminicli/export_test.go:361]: If `t.TempDir()` is used, the testing package will automatically remove the temporary directory when the test case finishes.

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-cli/geminicli/export_test.go (hunk 1 above)
Comment (wtsclwq) [reply app/ts-cli/geminicli/export_test.go]: refactored.

File: app/ts-cli/geminicli/export_test.go (hunk 2 above)
Comment (wtsclwq) [reply app/ts-cli/geminicli/export_test.go]: refactored

File: app/ts-cli/geminicli/export.go [hunk 3]
```diff
@@ -0,0 +1,1001 @@
...
+limitations under the License.
+*/
+
+package geminicli
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"encoding/binary"
```
Comment (scuzyj) [inline app/ts-cli/geminicli/export.go:23]: The sequence of go reference packages needs to be adjusted, reference the standard library and then the third-party library

File: app/ts-cli/geminicli/export.go [hunk 4]
```diff
@@ -0,0 +1,1001 @@
...
+				return err
+			}
+		}
+		fmt.Fprintf(outputWriter, "\n")
+	}
+	return nil
+}
+
+// writSingleTsspFile writes a single tssp file's all records.
+func (e *Exporter) writSingleTsspFile(filePath string, outputWriter io.Writer, index *tsi.MergeSetIndex) error {
```
Comment (scuzyj) [inline app/ts-cli/geminicli/export.go:649]: The word is misspelled. It should be write, not writ

File: app/ts-cli/geminicli/export.go [hunk 5]
```diff
@@ -0,0 +1,1001 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+// writeSingleRecord parses a record and a series key to line protocol, and writes it.
+func (e *Exporter) writeSingleRecord(outputWriter io.Writer, seriesKey [][]byte, rec record.Record) error {
+	tm := rec.Times()[0]
+	if !e.filter.Filter(tm) {
```
Comment (scuzyj) [inline app/ts-cli/geminicli/export.go:722]: Should we judge the time range, not only the minimum time?

File: app/ts-cli/geminicli/export.go [hunk 6]
```diff
@@ -0,0 +1,1001 @@
...
+	}
+	return nil
+}
+
+// writeSingleRow parse a single row to lint protocol, and writes it.
+func (e *Exporter) writeSingleRow(row influx.Row, outputWriter io.Writer) error {
+	measurementWithVersion := row.Name
+	var measurementName string
+	measurementName = influx.GetOriginMstName(measurementWithVersion)
+	measurementName = EscapeMstName(measurementName)
```
Comment (scuzyj) [inline app/ts-cli/geminicli/export.go:880]: Consider the reuse of frequent memory applications.

File: app/ts-cli/geminicli/export.go (hunk 5 above)
Comment (wtsclwq) [reply app/ts-cli/geminicli/export.go:722]: I filtered the maximum and minimum times in `writeSingleTsspFile()`, but did not encapsulate them as methods.
```go
if rec.MaxTime(true) < e.filter.startTime || rec.MinTime(true) > e.filter.endTime {
	continue
}
``` 
In the latest commit, I encapsulated two new methods

File: app/ts-cli/geminicli/export.go (hunk 6 above)
Comment (wtsclwq) [reply app/ts-cli/geminicli/export.go:880]: In the leaste commit, I tried it on `writeSingleRecord()` and  `writeSingleRow()`

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-cli/geminicli/export.go [hunk 7]
```diff
@@ -0,0 +1,1054 @@
...
+			if rpDir.IsDir() {
+				rpPath := path.Join(ptWalPath, rpDir.Name())
+				d.rpToWalDirMap[ptWithRp] = rpPath
+			}
+		}
+	}
+	return nil
+}
+
+const stdOtuMark = "-"
```
Comment (huang-feiteng) [inline app/ts-cli/geminicli/export.go:208]: stdOutMark?

File: app/ts-cli/geminicli/export.go [hunk 8]
```diff
@@ -0,0 +1,1054 @@
...
+	e.databases = clc.Databases
+	e.retentions = clc.Retentions
+	e.outPutPath = clc.Out
+	// A filter to filter points by time.
+	// TODO(wtsclwq) Support filter measurement.
+	e.filter = NewLineFilter()
+	e.compress = clc.Compress
+
+	// If output fd is stdout.
+	if e.usingStdOut() {
```
Comment (huang-feiteng) [inline app/ts-cli/geminicli/export.go:347]: !e.usingStdOut?

File: app/ts-cli/geminicli/export_test.go [hunk 9]
```diff
@@ -0,0 +1,823 @@
+/*
+Copyright 2022 Huawei Cloud Computing Technologies Co., Ltd.
```
Comment (huang-feiteng) [inline app/ts-cli/geminicli/export_test.go:2]: 2022->2023

**Pull request closed without merging**